lib/router/src/cache.rs

```
@@ -1,426 +1,367 @@
-    ///
-    /// Returns a handle that may be used to store an ite,. If there is no available
-    /// capacity, idle entries may be evicted to create capacity.
-    ///
-    /// An error is returned if there is no available capacity.
-    pub fn reserve(&mut self) -> Result<Reserve<K, V, N>, CapacityExhausted> {
+    pub fn reserve(&mut self) -> Result<Reserve<K, V>, CapacityExhausted> {
         if self.vals.len() == self.capacity {
-            // Only whole seconds are used to determine whether a node should be retained.
-            // This is intended to prevent the need for repetitive reservations when
-            // entries are clustered in tight time ranges.
-            let max_age = self.max_idle_age.as_secs();
-            let now = self.now.now();
-            self.vals.retain(|_, n| {
-                let age = now - n.last_access();
-                age.as_secs() <= max_age
-            });
-
-            if self.vals.len() == self.capacity {
-                return Err(CapacityExhausted {
-                    capacity: self.capacity,
-                });
+            match self.expirations.poll() {
+                // The cache is at capacity but we are able to remove a value.
+                Ok(Async::Ready(Some(entry))) => {
+                    self.vals.remove(entry.get_ref());
+                }
+
+                // `Ready(None)` can only be returned when expirations is
+                // empty. We know `expirations` is not empty because `vals` is
+                // not empty.
+                Ok(Async::Ready(None)) => unreachable!(),
+
+                // The cache is at capacity and no values can be removed.
+                Ok(Async::NotReady) => {
+                    return Err(CapacityExhausted {
+                        capacity: self.capacity,
+                    });
+                }
+
+                Err(_e) => {
+                    return Err(CapacityExhausted {
```

I'm not sure if it's correct to return `CapacityExhausted` here --- the error would be a timer error in this case, and I think it would be confusing to return `CapacityExhausted` when that isn't _actually_ the problem. Instead we should either propagate this error or panic, depending on whether or not the situations where the timer can error should be considered fatal...

---

Yea good call. I think depending on what I just mentioned above about `Async::Ready(None)`, we may want to be able to handle errors other than `CapacityExhausted`.